app/src/main/java/org/mozilla/focus/utils/DownloadUtils.java

```diff
@@ -0,0 +1,199 @@
+/* -*- Mode: Java; c-basic-offset: 4; tab-width: 20; indent-tabs-mode: nil; -*-
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ * This file incorporates work covered by the following copyright and
+ * permission notice:
+ *
+ *   Copyright (C) 2006 The Android Open Source Project
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License. */
+
+package org.mozilla.focus.utils;
+
+import android.net.Uri;
+import android.support.annotation.Nullable;
+import android.webkit.MimeTypeMap;
+
+import org.mozilla.focus.web.Download;
+
+import java.io.ByteArrayOutputStream;
+import java.io.UnsupportedEncodingException;
+import java.util.Locale;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class DownloadUtils {
+
+    /**
+     * Guess the name of the file that should be downloaded.
+     *
+     * This method is largely identical to {@link android.webkit.URLUtil#guessFileName}
+     * which unfortunately does not implement RfC 5987.
+     *
+     * @param download the Download object containing information about the request
+     * @return file name including extension
+     */
+    public static String guessFileName(Download download) {
```

Comment:
\[PMD\] INFO: The method 'guessFileName' has a Modified Cyclomatic Complexity of 20\.